### PR TITLE
[MDB IGNORE] Updates the library's machines to TGUI

### PR DESCRIPTION
**Pull request by LemonInTheDark**
### Overarching behavior stuff:

Updates all three library machines to tgui. So that's the visitor console, the book management console, and the scanner.
Adds some helper datums to make passing book information around easier
Adds hard limits on book title/author/contents
Reworks how bookshelves load books in. Instead of loading lazily, we load them during init in a separate subsystem. This opens us up to being able to do fun things, like autofilling a library console's inventory page. 
Adds a map config option for what areas count as part of "a library" in the eyes of the above system so kilo isn't woefully underfilled.
You can now choose what type of poster to print in the console, rather then getting a random item.
You can now choose to heed the warnings of the eldritch software company.
You can see what type of bible the chaplin has selected/who he worships?
You can no longer infinitely dump books into the library scanner.

### Implementation details, BYOND:

Any input that makes a db request will now A: freeze up any other db inputs until it's finished, and B: Start a 1 second timer before any new db requests can be made. You're also not allowed to make the same search request twice.
Of note, I'm handling html encoding in a very targeted way.
All book_data datums need to have html encoded values. get_title/author/content exist so a defaulting and tgui appropriate version can be loaded in. This somewhat matches with the trusted var on set_title, it exists to prevent double html encoding.

Looks like this
Input/DB (Book data should be html encoded)
Inside book datum (Book data should be html encoded)
Sending to tgui (Book data should be decoded during extraction with the get_() procs)
Sending anywhere else (Book data should be html encoded, otherwise it's an xss vuln)

### Implementation details, TGUI:

The visitor and book management console share the same data/act pipeline, which is why they're parented/subtyped
They also share a page selection component, which is why the visitor's console imports it.
I'm using a custom theme for emag visuals
The page selection component is kinda cursed, the left and right controls are fine, but I'm trying to get a `<< < [page/max] > >>` setup going, and that means resetting the center input after every change so the placeholder can be used
This ends up being slightly hacky. I'm sorry.

I've added a custom theme for the library console when it's emagged. There's a few semi cursed things about this.
1: I'm overriding a name to color mapping list in an odd way, since I'm pretty sure it's precomputed. Might not be needed though, I'm not the most comfortable with scss.
2: I added a new svg background for the ui, because I thought it would be a nice touch. I'm very out of my depth with this. It's currently presenting two problems.
One, is it even worth having in the first place. I don't have a very good concept of cost for uis like this, so if any part of this is needless bloat I'll just nuke it.
Two, how the hell do you compress these things. I've tried online options, and they'll display in my editor, but when I try and use them in the theme itself they just crap out. It's called -raw for a reason, I realize the current size is nowhere near acceptable. On that note, may redo the thing with less spirals.
As a part of all that I've added a dimness property to dimmer scss so themes can control how dark things get. Need this cause otherwise things just look washed out when modals are displayed.
 
Oh also, I implemented a custom tab setup for this ui. I have no idea why it was literally like 5 months ago.
I think it looks pretty nice, but if you want me to nuke it I can. Sorry for any headache around this.

One more thing.
For one of the pages in the ui, I'm displaying html raw. This is so markdownified books look goodish. 
I'm sanitizing them of course, but I ran into what I think are holes in our current whitelist. I think these are fine, can't think of any problems extra bold, strikethrough, underline and accent could cause, but I'm me so I wait to be proved an idiot.

Sorry if I missed anything here, this has been a few weeks of nightcode and I'm kinda running on fumes at the moment
### Pictures 
Library Scanner
image
Visitor's Console
image
Inventory Page
image
Checkout Manager
image
Checkout Popup
image
Archive/DB search
image
Upload to Newscaster/DB + Scanner Cache preview
image
Upload to DB popup
image

Misc Printing, Bibles and Posters
image
Collapsed Tabs
image
Emag Theme + Forbidden Page
image
Scanner Cache preview with html tags
image

## Why It's Good For The Game

Haha user interface go brrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrr
TODO:

- [x] Audit input box length, see if you can increase them to better match the size limits on titles and author pages
- [x] Check to see what markdown required tags exist, add them
- [ ] Lower the size of that background
- [x] Commit and push the remaining map repaths, lost in the sauce of my shitty pc
- [ ] Don't kill stylemistake
- [x] Default Bible Name
- [x] Test print page with max length everything, make sure it's ok

## Changelog

:cl:
expansion: The library computers have received a graphical and functional upgrade. Better inventory management, poster printing, bible viewing, and well, TGUI.
server: Bookcases will generate way more db calls during init then is normal. If this causes issues please bug me and I'll work out a solution
/:cl:

### CODE REVIEW

**Comment by JohnFulpWillard** [comment]
YES

**Comment by ShizCalev** [comment]
Conflicts.

File: tgui/packages/tgui/interfaces/LibraryScanner.js [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    <Window
+      title="Library Scanner"
+      width={350}
+      height={150}>
+      <BookScanning />
+    </Window>
+  );
+};
+
+export const BookScanning = (props, context) => {
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/LibraryScanner.js:16]:
```suggestion
const BookScanning = (props, context) => {
```
There is no need to export these components if you're only using them here

File: tgui/packages/tgui/interfaces/LibraryVisitor.js [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    <Window
+      title="Library Lookup Console"
+      width={702}
+      height={421}>
+      <BookListing />
+    </Window>
+  );
+};
+
+export const BookListing = (props, context) => {
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/LibraryVisitor.js:19]:
```suggestion
const BookListing = (props, context) => {
```
same here

File: tgui/packages/tgui/interfaces/LibraryVisitor.js [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+          disabled={!can_db_request}
+          call_on_change={(value) => act('switch_page', {
+            page: value,
+          })} />
+      </Stack.Item>
+    </Stack>
+  );
+};
+
+export const SearchAndDisplay = (props, context) => {
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/LibraryVisitor.js:62]:
```suggestion
const SearchAndDisplay = (props, context) => {
```
and here

File: tgui/packages/tgui/interfaces/LibraryScanner.js [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+  } = data;
+  if (!has_book && !has_cache) {
+    return (
+      <NoticeBox>
+        Insert a book to scan
+      </NoticeBox>
+    );
+  }
+  return (
+    <Flex
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/LibraryScanner.js:31]: Could you use ```Stack``` here instead of flex?

Comment (LemonInTheDark) [reply tgui/packages/tgui/interfaces/LibraryScanner.js:31]: Yes 100%, just never noticed

File: tgui/packages/tgui/styles/themes/spookyconsole.scss [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+@use 'sass:meta';
+@use 'sass:map';
+
+@use '../colors.scss' with (
+  $primary: #3f021a,
+  $good: #010005,
+  $bad: #970934,
+  $fg-map-keys: (),
+  $bg-map-keys: (),
+);
```
Comment (stylemistake) [inline tgui/packages/tgui/styles/themes/spookyconsole.scss:11]: Bruh, not another theme...

Comment (LemonInTheDark) [reply tgui/packages/tgui/styles/themes/spookyconsole.scss:11]: I'm so sorry

(It's genuinely really niche, if it's too much overhead we could cut it. I did it cause that's what came to mind when picturing the ui, and cause I wanted to give the system a go)

**Comment by optimumtact** [comment]
lemon bit off more than she could chew

**Comment by JohnFulpWillard** [comment]
NOOOOOOOOOO

**Comment by LemonInTheDark** [comment]
I will return, soon

**Comment by LemonInTheDark** [comment]
I am inevitable

**Pull request closed without merging**